### ISSUE TITLE
Failing tests on Windows

### ISSUE THREAD

**Issue by thtrummer** [openmm/openmm#2734]
After fixing most of the minor problems, there are a few remaining tests on Windows which fail:

**TestCudaVirtualSitesSingle**
*Assertion failure at TestVirtualSites.h:425.  Expected -85.2619, found -84.3424*

**TestCudaAmoebaGeneralizedKirkwoodForceSingle**
*Assertion failure at TestCudaAmoebaGeneralizedKirkwoodForce.cpp:7040.  testGeneralizedKirkwoodVillinMutualPolarizationRelative difference in norms 1.07946e-05 larger than allowed tolerance at particle=519: norms=94.2079 expected norm=94.2089*

**TestCudaAmoebaGeneralizedKirkwoodForceMixed**
*Assertion failure at TestCudaAmoebaGeneralizedKirkwoodForce.cpp:7040.  testGeneralizedKirkwoodVillinMutualPolarizationRelative difference in norms 1.07946e-05 larger than allowed tolerance at particle=519: norms=94.2079 expected norm=94.2089*

**TestCudaHippoNonbondedForceSingle**
*Assertion failure at TestCudaHippoNonbondedForce.cpp:1505.   Expected [5.75845, -117.857, 39.0743], found [5.11468, -117.965, 39.0257]*

**TestCudaHippoNonbondedForceMixed**
*Assertion failure at TestCudaHippoNonbondedForce.cpp:1505.   Expected [5.75845, -117.857, 39.0743], found [5.11465, -117.965, 39.0257]*

Is this expected or is it worth looking into it?

**Comment by peastman** [comment]
I don't know of any reason those tests would behave differently on Windows than Linux.  We at least need to figure out what's going on with them.

It looks like there are two tests that (just barely) fail in single and mixed precision mode, but pass in double precision.  Those ones might well just be an overly strict tolerance.  TestCudaVirtualSites is the more concerning one.  Though again, it only fails in single precision mode, not mixed or double.  But I don't know why the OS would make any difference.  All the calculations are running on the GPU.

What kind of GPU are you testing with?

**Comment by thtrummer** [comment]
Funnily enough there is a difference on Windows. When using a debug build [CudaContext](https://github.com/openmm/openmm/blob/e4c1f73c6875452c7d00b0922419806333a4518c/platforms/cuda/src/CudaContext.cpp#L599) adds debug symbols for the device code as well (only on Windows though). This disables all optimizations which in combination with `--use_fast_math` seems to lead to the error. Removing the parameter makes the test pass again (it's actually sufficient to only remove it for `integrationUtilities.cc` but that might be accidental).

Unfortunately there is a lot of device code involved and I don't see an easy way to pin point the function that's numerically unstable compared to the full precision version (at least not without writing a test for each function).

```
 CUDA Device Query (Runtime API) version (CUDART static linking)

Detected 1 CUDA Capable device(s)

Device 0: "GeForce GTX 1070"
  CUDA Driver Version / Runtime Version          11.0 / 10.2
  CUDA Capability Major/Minor version number:    6.1
  Total amount of global memory:                 8192 MBytes (8589934592 bytes)
  (15) Multiprocessors, (128) CUDA Cores/MP:     1920 CUDA Cores
  GPU Max Clock rate:                            1772 MHz (1.77 GHz)
  Memory Clock rate:                             4004 Mhz
  Memory Bus Width:                              256-bit
  L2 Cache Size:                                 2097152 bytes
  Maximum Texture Dimension Size (x,y,z)         1D=(131072), 2D=(131072, 65536), 3D=(16384, 16384, 16384)
  Maximum Layered 1D Texture Size, (num) layers  1D=(32768), 2048 layers
  Maximum Layered 2D Texture Size, (num) layers  2D=(32768, 32768), 2048 layers
  Total amount of constant memory:               zu bytes
  Total amount of shared memory per block:       zu bytes
  Total number of registers available per block: 65536
  Warp size:                                     32
  Maximum number of threads per multiprocessor:  2048
  Maximum number of threads per block:           1024
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (2147483647, 65535, 65535)
  Maximum memory pitch:                          zu bytes
  Texture alignment:                             zu bytes
  Concurrent copy and kernel execution:          Yes with 2 copy engine(s)
  Run time limit on kernels:                     Yes
  Integrated GPU sharing Host Memory:            No
  Support host page-locked memory mapping:       Yes
  Alignment requirement for Surfaces:            Yes
  Device has ECC support:                        Disabled
  CUDA Device Driver Mode (TCC or WDDM):         WDDM (Windows Display Driver Model)
  Device supports Unified Addressing (UVA):      Yes
  Device supports Compute Preemption:            Yes
  Supports Cooperative Kernel Launch:            Yes
  Supports MultiDevice Co-op Kernel Launch:      No
  Device PCI Domain ID / Bus ID / location ID:   0 / 1 / 0
```

**Comment by peastman** [comment]
That's really curious.  Am I understanding this correctly?

- All optimizations + `use_fast_math` = accurate
- No optimizations = accurate
- No optimizations + `use_fast_math` = NOT accurate?

**Comment by thtrummer** [comment]
Pretty much. Also:

- From the four options that `use_fast_math` implies only (but both) ` --prec-div=false --prec-sqrt=false` are required to make it fail
- Forcing the grid and block sizes all to `1` (to have nicer `printf` output) makes it pass (which may hint at a function that does thread gathering work)
- Using ` --fmad=true` for the optimized version makes it fail (while this should only cause a minor loss in precision)

Also unrelated: there is some kernel code like `const mixed scale = dtVel/(mixed) 0x100000000`. When `mixed` is a `float`, shouldn't this loose precision since `0x100000000` requires 32 bit and a `float` only has 24 bits of precision (probably harmless though)?

**Comment by peastman** [comment]
Very strange.  I don't have a better solution than to just make sure we use a combination of flags for which all tests pass.

> When mixed is a float, shouldn't this loose precision since 0x100000000 requires 32 bit and a float only has 24 bits of precision (probably harmless though)?

That number can be exactly represented in single precision.  It's just 1*(2^32).  Single precision supports exponents up to 2^127.

**Comment by thtrummer** [comment]
Are you able to reproduce it? It would hate to change it for something that may very well be an idiosyncrasy of my system.

> That number can be exactly represented in single precision. It's just 1*(2^32). Single precision supports exponents up to 2^127.

It helps when looking at the binary representation of the number instead of decimal one... :)

**Comment by peastman** [comment]
I can't reproduce it on Linux.  I tried adding `-g -O0` to the optimization flags, but it still passed.  I don't have a Windows system with CUDA to test on.

**Comment by thtrummer** [comment]
In [this line](https://github.com/thtrummer/openmm/blob/8c43e37a887eeaa919f31c762c5ee8841bba9a22/platforms/cuda/src/CudaContext.cpp#L605) you have to add an uppercase [-G](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-altering-compiler-linker-behavior-device-debug) to have the same effect, right?

**Comment by peastman** [comment]
You're right, I was able to reproduce it with that option.

I narrowed things down a bit, since `--use_fast_math` implies several other options.  The particular combination of flags that triggers the problem is ` --prec-div=false --prec-sqrt=false -G`.  If you remove any one of those, it works.

I wonder if this could be a compiler bug triggered by giving it inconsistent options?  The documentation says that `-G` disables all optimizations.  So we tell it to disable all optimizations, but then also tell it to include some specific optimizations.  I could understand if it gets confused in that case.
